File: src/services/checker.rs
use crate::services::collector::Proxy;
use serde::Deserialize;
use tokio::sync::mpsc::Sender;

#[derive(Deserialize)]
struct HttpBinResponse {
    origin: String,
}

pub struct CheckProxyResponse {
    pub alive: bool,
    pub host: String,
    pub port: u16,
}

/**
 * checks if it is possible to use proxy
 */
async fn send_proxy_request(sender: Sender<CheckProxyResponse>, proxy: Proxy) {
    match reqwest::Client::builder()
        .proxy(
            reqwest::Proxy::http(format!("http://{}:{}", proxy.host, proxy.port))
                .expect("Cannot build proxy"),
        )
        .build()
    {
        Ok(client) => {
            if let Ok(response) = client.get("http://httpbin.org/ip").send().await {
                if let Ok(body) = response.json::<HttpBinResponse>().await {
                    sender
                        .try_send(CheckProxyResponse {
                            alive: body.origin.eq(&proxy.host),
                            host: proxy.host,
                            port: proxy.port,
                        })
                        .ok();
                }
            }
        }
        Err(e) => println!("Cannot build client: {}", e),
    }
}

pub fn check(sender: Sender<CheckProxyResponse>, proxies: Vec<Proxy>) {
    for proxy in proxies {
        tokio::task::spawn(send_proxy_request(sender.clone(), proxy));
    }
}

File: src/services/checker.ts
import { EventEmitter } from "events";
import { Client } from 'undici';
import { Proxy } from '../services/collector'

export async function check(emitter: EventEmitter, proxy: Proxy) {
  try {
    let client = new Client(`http://${proxy.host}:${proxy.port}`);
    let response = await client.request({
      path: 'http://httpbin.org/ip?json',
      method: 'GET'
    });
    response.body.setEncoding('utf8');
    let data = '';
    for await (const chunk of response.body) data += chunk;
    let obj = JSON.parse(data);
    emitter.emit('foundCheckedProxies', {
      alive: obj.origin === proxy.host,
      host: proxy.host,
      port: proxy.port
    })
  } catch(e) {

  }
}